README.md
flXHR & flensed

flXHR (and its parent project "flensed") is an open-source project using flash+javascript to create a drop-in replacement of the XMLHTTPRequest object for cross-domain Ajax.

The site and all project data used to live on its own site at http://flensed.com, but has now been moved/dumped here on github to ensure that the project remains open and available for posterity.

Given recent advances such as CORS (Cross-origin Resource Sharing) which implement cross-domain Ajax using the native XHR interface, flXHR has now accomplished its goal and is no longer necessary. It would only be useful in older browsers as a shim.

Moreover, given that flash has become increasingly marginalized in the web development arena, in favor of standards-based and open web technology choices such as HTML5 and JavaScript, this project is no longer being maintained. All content and code are hereby donated to the public domain (sans the MIT license).


releasenotes.md
flXHR 1.0.6 RELEASE NOTES

23 February 2011
Issue #21 fixed.
flXHR 1.0.5 RELEASE NOTES

15 January 2010
The release of jQuery 1.4 required some minor changes to how flXHR manages the public readyState property to keep flXHR working with jQuery via the jQuery.flXHRproxy plugin. If you are not using flXHR with jQuery, this update is probably not as relevant to you.
jQuery flXHRproxy plugin 1.2.2 RELEASE NOTES

25 August 2009
Reduced code size by a few bytes, in addressing a bug with Opera and the way it handled array manipulation. Now using jQuery's $.makeArray() utility function, which reliably works cross-browser.

Also, added an if-statement check to prevent calling the success handler if the jQuery framework call doesn't define one, like for instance if you use flXHR with the $.load() function which doesn't use a success handler.
flXHR 1.0.4 RELEASE NOTES

11 August 2009
This patch release addresses some minor behaviors of flXHR which are non-standard compared to the native XHR object, which created integration issues for some projects, including Strophe. The behaviors are related to the allowed values of 'readyState' property, and when the 'onreadystatechange' handler is called. In addition, the minified/compressed build-ready version of flXHR.js reduced ~0.6k, a 5% reduction!

'readyState' will now never publicly have a '-1' value (which it still keeps internally during the initial object construction phase while it queues API calls). 'onreadystatechange' will not be called on the '0' state as it was previously.
jQuery flXHRproxy plugin 1.2.1 RELEASE NOTES

10 July 2009
Better integration with the framework's error mechanism now. Preferred to use the "error" event in the $.ajax call rather than manually specifying "onerror" property of flXHR instance.

Also, improved the signature of the "success" handler. It now has a third parameter, which is the flXHR instance (XHR) object the success is in reference to. Helps identify requests and their responses to keep them in sync.
DojoX.io | XHR Plugins extension | flXHRproxy plugin 1.1 RELEASE NOTES

9 July 2009
Better integration with the framework's error mechanism now. Preferred to use the "error" event in the Xhr call rather than manually specifying "onerror" property of flXHR instance.
Prototype extension flXHRproxy plugin 1.1 RELEASE NOTES

9 July 2009
Better integration with the framework's error mechanism now. Preferred to use the "onFailure" and "onException" events rather than manually specifying "onerror" property of flXHR instance.
Mootools extension flXHRproxy plugin 1.0 RELEASE NOTES

8 July 2009
Initial release of the Mootools flXHRproxy plugin. Works the same as the other frameworks' plugins. Thanks to Zohaib (MaXPert) for helping get the initial code written!
flXHR 1.0.3.1 RELEASE NOTES

3 June 2009
This is a build-version release (sub patch level) since no flXHR code is changed, but the bundled CheckPlayer release was changed due to a CheckPlayer bug relating to auto upgrades failing to happen. This affected flXHR since the "autoUpdatePlayer" flag was essentially failing to operate correctly. All flXHR deployments should upgrade to get the latest CheckPlayer logic with the bug fixed, so that flXHR can upgrade player plugins when necessary/desired.
jQuery flXHRproxy plugin 1.1 RELEASE NOTES

2 June 2009
Issue #17 fixed.
Prototype extension flXHRproxy plugin 1.0.1 RELEASE NOTES

2 June 2009
Minor bug fix to code logic for url matching logic
Prototype extension flXHRproxy plugin 1.0 RELEASE NOTES

22 May 2009
Initial public release of Prototype plugin which registers flXHR as a transport for use with cross-domain Ajax calls.
DojoX.io | XHR Plugins extension | flXHRproxy plugin 1.0 RELEASE NOTES

10 May 2009
Initial public release of Dojo plugin which registers flXHR as a transport for use with the DojoX.io | XHR Plugins extension External Link.
jQuery flXHRproxy plugin 1.0 RELEASE NOTES

7 March 2009
Initial public release of jQuery plugin which registers flXHR as a transport for use with the XHR Registry plugin External Link.
flXHR 1.0.3 RELEASE NOTES

7 March 2009
Issue #16 fixed.
Also, some minor internal changes for efficiency and robustness.
flXHR 1.0.2 RELEASE NOTES

16 January 2009
Issue #15 fixed.
An IE bug was discovered with jQuery 1.3 External Link (unrelated to flXHR) which requires a patch to make jQuery 1.3 work inside an iframe where the iframe domain and the main page domain are different.

The bug report can be found here: http://dev.jquery.com/ticket/3898 External Link
The patch can be found here: http://dev.jquery.com/changeset/6120 External Link
A patched version of jQuery can be found here: http://flxhr.flensed.com/js/jquery-1.3.js

jQuery 1.3.1 External Link has been released and works fine inside of iframes and with flXHR.
flXHR 1.0.1 RELEASE NOTES

14 January 2009
Issue #14 fixed.
Suggestion implemented.
A few minor bug fixes, and the new configurable "noCacheHeader" boolean flag option, which controls if a 'pragma:no-cache' request header is automatically appended to all requests (defaults to 'true', as before).
flXHR 1.0 RELEASE NOTES

19 December 2008
Issue #11 fixed.
NOTE: flXHR 1.0 is only compatible with flensedCore 1.0 (or higher) and CheckPlayer 1.0 (or higher).
Significant bug fixes, code optimizations, and code compression. Also, flXHR.swf can now be open-source compiled using the Flex SDK 'mxmlc' compiler.



code/releases/flensed-1.0/src/flCookie/flCookie.src.js
/*	flCookie 0.1 <http://flcookie.flensed.com/>
	Copyright (c) 2009 Kyle Simpson, Getify Solutions, Inc.
	This software is released under the MIT License <http://www.opensource.org/licenses/mit-license.php>

	====================================================================================================
*/

(function(global){
	// frequently used variable declarations, for optimized compression
	var oWIN = global,
		oDOC = global.document,
		bTRUE = true,
		bFALSE = false,
		sEMPTY = "",
		sUNDEF = "undefined",
		sOBJECT = "object",
		sFUNCTION = "function",
		sSTRING = "string",
		sDIV = "div",
		sONUNLOAD = "onunload",
		tmp = null,
		_flcookie_idc = 0,
		_flensed = null,
		_chkplyr = null,
		_cp_ins = null,
		js_type = "text/javascript",
		flCookie_js = "flCookie.js",			// SHOULD NOT rename the file or change this line
		flensed_js = "flensed.js",				// ditto
		checkplayer_js = "checkplayer.js",		// ditto
		flCookie_swf = "flCookie.swf",			// ditto
		this_script = flCookie_js,
		fPARSEINT = global.parseInt,
		fSETTIMEOUT = global.setTimeout,
		fCLEARTIMEOUT = global.clearTimeout
	;

	if (typeof global.flensed === sUNDEF) { global.flensed = {}; }
	if (typeof global.flensed.flCookie !== sUNDEF) { return; }	// flCookie already defined, so quit
	
	_flensed = global.flensed;
	
	fSETTIMEOUT(function() {
		var base_path_known = bFALSE,
			scriptArry = oDOC.getElementsByTagName("script"),
			oHEAD = oDOC.getElementsByTagName("head")[0],
			scrlen = scriptArry.length;
		try { _flensed.base_path.toLowerCase(); base_path_known = bTRUE; } catch(err) { _flensed.base_path = sEMPTY; }
	
		function load_script(src,type) {
			for (var k=0; k<scrlen; k++) {
				if (typeof scriptArry[k].src !== sUNDEF) {
					if (scriptArry[k].src.indexOf(src) >= 0) { break; }  // this script already loaded/loading...
				}
			}
			var scriptElem = oDOC.createElement("script");
			scriptElem.setAttribute("src",_flensed.base_path+src);
			if (typeof type !== sUNDEF) { scriptElem.setAttribute("type",type); }
			oHEAD.appendChild(scriptElem);
		}
		
		if ((typeof scriptArry !== sUNDEF) && (scriptArry !== null)) {
			if (!base_path_known) {
				var idx=0;
				for (var k=0; k<scrlen; k++) {
					if (typeof scriptArry[k].src !== sUNDEF) {
						if (((idx=scriptArry[k].src.indexOf(flensed_js)) >= 0) || ((idx=scriptArry[k].src.indexOf(flCookie_js)) >= 0)) {
							_flensed.base_path = scriptArry[k].src.substr(0,idx);
							break;
						}
					}
				}
			}
		}
		try { _flensed.checkplayer.module_ready(); } catch (err2) { load_script(checkplayer_js,js_type); }
		load_script(flensed_js,js_type);
	
		var coreInterval = null;
		(function waitForCore() {
			try { _flensed.ua.pv.join("."); } catch (err) { coreInterval = fSETTIMEOUT(arguments.callee,25); return; }
			_flensed.bindEvent(oWIN,sONUNLOAD,function(){
				try {
					global.flensed.unbindEvent(oWIN,sONUNLOAD,arguments.callee);
					for (var k in _flcookie) {
						if (_flcookie.hasOwnProperty(k)) {
							try { _flcookie[k] = null; } catch (err2) { }
						}
					}
					_flensed.flCookie = _flcookie = _flensed = _cp_ins = _chkplyr = null;
				}
				catch (err3) { }
			});
		})();
		function clearCoreInterval() { fCLEARTIMEOUT(coreInterval); try { oWIN.detachEvent(sONUNLOAD,clearCoreInterval); } catch (err) {} }
		if (coreInterval !== null) { try { oWIN.attachEvent(sONUNLOAD,clearCoreInterval); } catch(err3) {} }

		var dependencyTimeout = null;
		function clearDependencyTimeout() { fCLEARTIMEOUT(dependencyTimeout); try { oWIN.detachEvent(sONUNLOAD,clearDependencyTimeout); } catch (err) {} }
		try { oWIN.attachEvent(sONUNLOAD,clearDependencyTimeout); } catch (err4) {}
		dependencyTimeout = fSETTIMEOUT(function(){
			clearDependencyTimeout();
			try { 
				_flensed.checkplayer.module_ready(); 
			} catch (err2) { throw new global.Error("flCookie dependencies failed to load."); }
		},20000);	// only wait 20 secs max for CheckPlayer to load
	},0);

	_flensed.flCookie = function(cookieLocation,cookieNameSuffix,readyCB,errorCB) {
		if (typeof cookieLocation === sSTRING) {
			if (cookieLocation.length > 0 && cookieLocation.charAt(cookieLocation.length-1) !== "/") cookieLocation += "/";
		}
		else cookieLocation = sEMPTY;
		if (typeof cookieNameSuffix !== sSTRING) {
			cookieNameSuffix = sEMPTY;
		}
		if (typeof readyCB !== sFUNCTION) readyCB = function(){};
		
		// Private Properties
		var	idNumber = ++_flcookie_idc,
			publicAPI,
			appendTo = null,
			constructInterval,
			styleClass = "flCookieHideSwf",
			_flcookie_css = bFALSE,
			_ready = bFALSE,
			_error = bFALSE,
			cookieId = sEMPTY,
			cookieObj,
			oBODY = oDOC.getElementsByTagName("body"),
	
			autoUpdatePlayer = bFALSE, 	// TODO: hook up configuration to allow author to set these at instantiation time
			appendToId = null,
			cookieIdPrefix = "flCookie_swf"
		;
	
		// Private Methods
		var constructor = function() {
			cookieId = cookieIdPrefix+"_"+idNumber;
			
			function clearConstructInterval() { fCLEARTIMEOUT(constructInterval); try { oWIN.detachEvent(sONUNLOAD,clearConstructInterval); } catch (err) { } }
			try { oWIN.attachEvent(sONUNLOAD,clearConstructInterval); } catch (err) { }	// only IEoWIN would leak memory this way
			(function waitForCore() {
				try { _flensed.bindEvent(oWIN,sONUNLOAD,destructor); } catch (err) { constructInterval = fSETTIMEOUT(arguments.callee,25); return; }
				clearConstructInterval();
				constructInterval = fSETTIMEOUT(continueConstructor,1);
			})();
		}();
			
		function continueConstructor() {
			if (appendToId === null) { appendTo = oBODY[0]; }
			else { appendTo = _flensed.getObjectById(appendToId); }
			
			try { appendTo.nodeName.toLowerCase(); _flensed.checkplayer.module_ready(); _chkplyr = _flensed.checkplayer; } catch (err) {	// make sure DOM object and checkplayer are ready
				// maybe set a timeout here in case the DOM obj (appendTo) never gets ready?
				constructInterval = fSETTIMEOUT(continueConstructor,25);
				return;
			}
			
			_flensed.bindEvent(oWIN,sONUNLOAD,destructor);
			_chkplyr = _flensed.checkplayer;
						
			if ((_cp_ins === null) && (typeof _chkplyr._ins === sUNDEF)) {
				try {
					_cp_ins = new _chkplyr(_flcookie.MIN_PLAYER_VERSION,checkCallback,bFALSE,updateCallback);
				}
				catch (err2) { doError(_flcookie.DEPENDENCY_ERROR,"flCookie: checkplayer Init Failed","The initialization of the 'checkplayer' library failed to complete."); return; }
			}
			else {
				_cp_ins = _chkplyr._ins;
				stillContinueConstructor();
			}
		}
	
		function stillContinueConstructor() {
			if (_flcookie_css === null && appendToId === null) {	// only if CSS hasn't been defined yet, and if flCookie is being added to the BODY of the page
				_flensed.createCSS("."+styleClass,"left:-1px;top:0px;width:1px;height:1px;position:absolute;");	// CSS to hide any flCookie instances added automatically to the BODY
				_flcookie_css = bTRUE;
			}

			var holder = oDOC.createElement(sDIV);
			holder.id = cookieId;
			holder.className = styleClass;
			appendTo.appendChild(holder);
			appendTo = null;
	
			var flashvars = {},
				params = { allowScriptAccess:"always" },
				attributes = { id:cookieId, name:cookieId, styleclass:styleClass },
				optionsObj = { swfCB:initSWF, swfEICheck:"setId" }
			;
	
			try {
				_cp_ins.DoSWF(cookieLocation+flCookie_swf, cookieId, "1", "1", flashvars, params, attributes, optionsObj);
			}
			catch (err2) { doError(_flcookie.DEPENDENCY_ERROR,"flCookie: checkplayer Call Failed","A call to the 'checkplayer' library failed to complete."); return; }
		}
	
		function checkCallback(checkObj) {
			if (checkObj.checkPassed) {
				stillContinueConstructor();
			}
			else if (!autoUpdatePlayer) {
				doError(_flcookie.PLAYER_VERSION_ERROR,"flCookie: Insufficient Flash Player Version","The Flash Player was either not detected, or the detected version ("+checkObj.playerVersionDetected+") was not at least the minimum version ("+_flcookie.MIN_PLAYER_VERSION+") needed by the 'flCookie' library.");
			}
			else {
				_cp_ins.UpdatePlayer();
			}
		}

		function updateCallback(checkObj) {
			if (checkObj.updateStatus === _chkplyr.UPDATE_CANCELED) {
				doError(_flcookie.PLAYER_VERSION_ERROR,"flCookie: Flash Player Update Canceled","The Flash Player was not updated.");
			}
			else if (checkObj.updateStatus === _chkplyr.UPDATE_FAILED) {
				doError(_flcookie.PLAYER_VERSION_ERROR,"flCookie: Flash Player Update Failed","The Flash Player was either not detected or could not be updated.");
			}
		}

		function initSWF(loadStatus) {
			//console.log("initSWF-1");
			if (loadStatus.status !== _chkplyr.SWF_EI_READY) { return; }
			//console.log("initSWF-2");

			clearIntervals();
			cookieObj = _flensed.getObjectById(cookieId);
			cookieObj.setId(cookieId);

			cookieObj.doOnError = doError;
			cookieObj.doOnReady = doReady;
			
			cookieObj.initCookie(cookieNameSuffix);
		}
		
		function finishConstructor(loadStatus) {
			//console.log("finishConstructor");
			_ready = bTRUE;
			fSETTIMEOUT(function(){
				try { readyCB(publicAPI); } catch (err) { doError(_flcookie.HANDLER_ERROR,"flCookie::readyCallback(): Error","An error occurred in the handler function. ("+err.message+")"); return; }
			},0);
		}
		
		function destructor() {
			try { global.flensed.unbindEvent(oWIN,sONUNLOAD,destructor); } catch (err) { }
			try {
				for (var j in publicAPI) {
					if (publicAPI.hasOwnProperty(j)) {
						try { publicAPI[j] = null; } catch (err3) { }
					}
				}
			}
			catch (err4) { }
			publicAPI = null;
	
			clearIntervals();
			if ((typeof cookieObj !== sUNDEF) && (cookieObj !== null)) {
				try { cookieObj.doOnError = null; doOnError = null; } catch (err5) { }
				try { cookieObj.doOnReady = null; doOnReady = null; } catch (err6) { }
				cookieObj = null;
			
				try { global.swfobject.removeSWF(cookieId); } catch(err10) { }
			}
			
			readyCB = null;
			errorCB = null;
		}

		function doReady() {
			//console.log("doReady");
			if (!_ready && !_error) finishConstructor();
		}
	
		function doError() {
			clearIntervals();
			_error = bTRUE;
			var errorObj;
			try {
				errorObj = new _flensed.error(arguments[0],arguments[1],arguments[2],publicAPI);
			}
			catch (err) {
				function ErrorObjTemplate() { this.number=0;this.name="flCookie Error: Unknown";this.description="Unknown error from 'flCookie' library.";this.message=this.description;this.srcElement=publicAPI;var a=this.number,b=this.name,c=this.description;function toString() { return a+", "+b+", "+c; } this.toString=toString; }
				errorObj = new ErrorObjTemplate();
			}
			var handled = bFALSE;
			try { 
				if (typeof errorCB === sFUNCTION) { errorCB(errorObj); handled = bTRUE; }
			}
			catch (err2) { 
				var prevError = errorObj.toString();
				function ErrorObjTemplate2() { this.number=_flcookie.HANDLER_ERROR;this.name="flCookie::errorCallback(): Error";this.description="An error occured in the handler function. ("+err2.message+")\nPrevious:["+prevError+"]";this.message=this.description;this.srcElement=publicAPI;var a=this.number,b=this.name,c=this.description;function toString() { return a+", "+b+", "+c; } this.toString=toString; }
				errorObj = new ErrorObjTemplate2();
			}
	
			if (!handled) {
				SETTIMEOUT(function() { _flensed.throwUnhandledError(errorObj.toString()); },1);
			}
		}

		function clearIntervals() {
			fCLEARTIMEOUT(constructInterval);
			constructInterval = null;
		}
		
		function getValue(name) {
			if (!_error) {
				if (typeof name !== sSTRING || name.length < 1) return null;
				var now = new Date().getTime(), parts, tmp;
				try { parts = cookieObj.getValue(name); } catch (err) {
					doError(_flcookie.CALL_ERROR,"flCookie::getValue(): Failed","The getValue() call failed to complete.");
					return null;
				}
				if (parts !== null) {
					parts = parts.match(/^([^\;]+);(.*)/m);
					if (parts.length !== 3 || (parts[1] !== "." && now >= fPARSEINT(parts[1]))) { // cookie is invalid or value has expired
						deleteValue(name);
						return null;
					}
					return parts[2];
				}
				else return null;
			}
			else return null;
		}
		
		function setValue(name,val,expires) {
			if (!_error) {
				if (typeof name !== sSTRING || name.length < 1 || typeof val !== sSTRING || val.length < 1) return null;
				if (typeof expires !== sUNDEF) {
					if (typeof expires !== sSTRING) expires = sEMPTY + expires;
					if (expires !== "." && expires.match(/[^0-9]/g)) expires = Date.parse(expires);
					else if (expires === sEMPTY) expires = ".";
				}
				else expires = ".";
				try { return cookieObj.setValue(name,expires+";"+val); } catch (err) {
					doError(_flcookie.IO_ERROR,"flCookie::setValue(): Failed","The setValue() call failed to complete.");
					return bFALSE;
				}
			}
			else return bFALSE;
		}
		
		function deleteValue(name) {
			if (!_error) {
				if (typeof name !== sSTRING || name.length < 1) return bFALSE;
				try { return cookieObj.deleteValue(name); } catch (err) {
					doError(_flcookie.IO_ERROR,"flCookie::deleteValue(): Failed","The deleteValue() call failed to complete.");
					return bFALSE;
				}
			}
			else return bFALSE;
		}		
	
		// Public API
		publicAPI = {
			instanceId:cookieId,
			ready:function(){return _ready;},
			getValue:getValue,
			setValue:setValue,
			deleteValue:deleteValue,
			Destroy:destructor
		};
		return publicAPI;
	};
	_flcookie = _flensed.flCookie;	// frequently used variable declarations
	
	// Static Properties
	_flcookie.HANDLER_ERROR = 10;
	_flcookie.CALL_ERROR = 11;
	_flcookie.DEPENDENCY_ERROR = 13;
	_flcookie.PLAYER_VERSION_ERROR = 14;
	_flcookie.SECURITY_ERROR = 15;
	_flcookie.IO_ERROR = 16;
	_flcookie.MIN_PLAYER_VERSION = "9";
	_flcookie.module_ready = function(){};
	

})(window);
